### ISSUE TITLE
azurerm_mssql_virtual_machine attend to recreate SQL VM resource if auto_backup settings is defined

### ISSUE THREAD

**Issue by arestarh** [terraform-providers/terraform-provider-azurerm#10604]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
**terraform** version: **0.14.3**
**AzureRM** Provider version: **2.47.0**

### Affected Resource(s)

* `azurerm_mssql_virtual_machine `

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_virtual_machine" "db_vm" {
  virtual_machine_id               = azurerm_windows_virtual_machine.db_vm.id
  sql_license_type                 = "PAYG"

  auto_patching {
    day_of_week                            = "Saturday"
    maintenance_window_duration_in_minutes = 60
    maintenance_window_starting_hour       = 6
  }

  auto_backup {
    encryption_enabled              = false
    retention_period_in_days        = 14
    system_databases_backup_enabled = true
    storage_blob_endpoint           = azurerm_storage_account.backup_st.primary_blob_endpoint
    storage_account_access_key      = azurerm_storage_account.backup_st.primary_access_key

    manual_schedule {
      full_backup_frequency           = "Daily"
      full_backup_start_hour          = 4
      full_backup_window_in_hours     = 4
      log_backup_frequency_in_minutes = 60
    }
  }
}

```

### Expected Behaviour
**terraform apply** create SQL VM resource during first run and will not recreate SQL VM resource during following **terraform apply** execution (when auto_backup settings is defined).

### Actual Behaviour
**terraform apply** failed with timeout (while SQL VM resource was actually created) and **terraform apply** attended to recreate SQL VM resource during following **terraform apply** execution (when auto_backup settings is defined).

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

**Comment by arestarh** [comment]
I'll try to describe several scenarios we've tried.

In AzureRM provider version 2.47 ability to configure Auto Backup for SQL VM was added.

1. Before we had started to use AzureRM provider version 2.47 we've configured Auto Backup for SQL VM manually via Azure Portal and **auto_backup** block was not added to terraform code. When I've tried to run **terraform plan** I've got next results:
```
-/+ resource "azurerm_mssql_virtual_machine" "db_vm" {
2021-02-15T13:04:23.3011014Z       ~ id                    = "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db" -> (known after apply)
2021-02-15T13:04:23.3011558Z       - r_services_enabled    = false -> null
2021-02-15T13:04:23.3015558Z         # (4 unchanged attributes hidden)
2021-02-15T13:04:23.3016483Z 
2021-02-15T13:04:23.3016825Z       - auto_backup { # forces replacement
2021-02-15T13:04:23.3017351Z           - encryption_enabled              = false -> null
2021-02-15T13:04:23.3019797Z           - retention_period_in_days        = 14 -> null
2021-02-15T13:04:23.3020270Z           - system_databases_backup_enabled = true -> null
2021-02-15T13:04:23.3020502Z 
2021-02-15T13:04:23.3020735Z           - manual_schedule {
2021-02-15T13:04:23.3021085Z               - full_backup_frequency           = "DAILY" -> null
2021-02-15T13:04:23.3021490Z               - full_backup_start_hour          = 4 -> null
2021-02-15T13:04:23.3021885Z               - full_backup_window_in_hours     = 4 -> null
2021-02-15T13:04:23.3022259Z               - log_backup_frequency_in_minutes = 0 -> null
2021-02-15T13:04:23.3022589Z             }
2021-02-15T13:04:23.3022788Z         }
2021-02-15T13:04:23.3022971Z 
2021-02-15T13:04:23.3023223Z         # (1 unchanged block hidden)
2021-02-15T13:04:23.3023492Z     }
2021-02-15T13:04:23.3023630Z 
```
After that I've changed azurerm_mssql_virtual_machine deinition according to settings we've configured in Azure Portal:
```
resource "azurerm_mssql_virtual_machine" "db_vm" {
  virtual_machine_id               = azurerm_windows_virtual_machine.db_vm.id
  sql_license_type                 = "PAYG"

  auto_patching {
    day_of_week                            = "Saturday"
    maintenance_window_duration_in_minutes = 60
    maintenance_window_starting_hour       = 6
  }

  auto_backup {
    encryption_enabled              = false
    retention_period_in_days        = 14
    system_databases_backup_enabled = true
    storage_blob_endpoint           = azurerm_storage_account.backup_st.primary_blob_endpoint
    storage_account_access_key      = azurerm_storage_account.backup_st.primary_access_key

    manual_schedule {
      full_backup_frequency           = "Daily"
      full_backup_start_hour          = 4
      full_backup_window_in_hours     = 4
      log_backup_frequency_in_minutes = 60
    }
  }
}
```
and run terrafrom plan again:
```
~ resource "azurerm_mssql_virtual_machine" "db_vm" {
2021-02-15T14:00:00.5313691Z         id                    = "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db"
2021-02-15T14:00:00.5316466Z         # (5 unchanged attributes hidden)
2021-02-15T14:00:00.5316906Z 
2021-02-15T14:00:00.5317194Z       ~ auto_backup {
2021-02-15T14:00:00.5318746Z           + storage_account_access_key      = "Key"
2021-02-15T14:00:00.5320016Z           + storage_blob_endpoint           = "blob_endpoint_url"
2021-02-15T14:00:00.5320889Z             # (3 unchanged attributes hidden)
2021-02-15T14:00:00.5321133Z 
2021-02-15T14:00:00.5321360Z           ~ manual_schedule {
2021-02-15T14:00:00.5323688Z               ~ log_backup_frequency_in_minutes = 0 -> 60
2021-02-15T14:00:00.5324179Z                 # (3 unchanged attributes hidden)
2021-02-15T14:00:00.5324491Z             }
2021-02-15T14:00:00.5324773Z         }
2021-02-15T14:00:00.5324988Z 
2021-02-15T14:00:00.5325284Z         # (1 unchanged block hidden)
2021-02-15T14:00:00.5325604Z     }
```
the following terraform apply failed with error:

> 2021-02-15T15:07:22.7626137Z Error: waiting for SQL Virtual Machine "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db" AutoBackupSettings to take effect: sqlvirtualmachine.SQLVirtualMachinesClient#Get: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded

and run terrafrom plan again:
```
~ resource "azurerm_mssql_virtual_machine" "db_vm" {
2021-02-15T15:11:33.8733780Z         id                    = "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db"
2021-02-15T15:11:33.8736742Z         # (5 unchanged attributes hidden)
2021-02-15T15:11:33.8737014Z 
2021-02-15T15:11:33.8737256Z       ~ auto_backup {
2021-02-15T15:11:33.8737559Z             # (5 unchanged attributes hidden)
2021-02-15T15:11:33.8737874Z 
2021-02-15T15:11:33.8738216Z           ~ manual_schedule {
2021-02-15T15:11:33.8739811Z               ~ log_backup_frequency_in_minutes = 0 -> 60
2021-02-15T15:11:33.8740527Z                 # (3 unchanged attributes hidden)
2021-02-15T15:11:33.8740908Z             }
2021-02-15T15:11:33.8741161Z         }
2021-02-15T15:11:33.8741327Z 
2021-02-15T15:11:33.8741979Z         # (1 unchanged block hidden)
2021-02-15T15:11:33.8742250Z     }
```
the following terraform apply failed with error:

> 2021-02-15T15:07:22.7626137Z Error: waiting for SQL Virtual Machine "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db" AutoBackupSettings to take effect: sqlvirtualmachine.SQLVirtualMachinesClient#Get: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded

2. Than I've manually deleted mentioned SQL VM resource and run terraform plan and apply, but received the same error after timeout exceeded (while SQL VM resource was actually created):

> 2021-02-15T15:07:22.7626137Z Error: waiting for SQL Virtual Machine "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db" AutoBackupSettings to take effect: sqlvirtualmachine.SQLVirtualMachinesClient#Get: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
 
I've run run terrafrom plan again:
```
2021-02-16T13:09:41.1573870Z   # azurerm_mssql_virtual_machine.db_vm is tainted, so must be replaced
2021-02-16T13:09:41.1574697Z -/+ resource "azurerm_mssql_virtual_machine" "db_vm" {
2021-02-16T13:09:41.1575889Z       ~ id                    = "/subscriptions/***/resourceGroups/test-rg/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/vm-db" -> (known after apply)
2021-02-16T13:09:41.1576906Z       - r_services_enabled    = false -> null
2021-02-16T13:09:41.1580762Z         # (4 unchanged attributes hidden)
2021-02-16T13:09:41.1581363Z 
2021-02-16T13:09:41.1630290Z       ~ auto_backup {
2021-02-16T13:09:41.1630673Z             # (5 unchanged attributes hidden)
2021-02-16T13:09:41.1630923Z 
2021-02-16T13:09:41.1631166Z           ~ manual_schedule {
2021-02-16T13:09:41.1632035Z               ~ full_backup_frequency           = "DAILY" -> "Daily"
2021-02-16T13:09:41.1655887Z               ~ log_backup_frequency_in_minutes = 0 -> 60
2021-02-16T13:09:41.1656477Z                 # (2 unchanged attributes hidden)
2021-02-16T13:09:41.1656843Z             }
2021-02-16T13:09:41.1657086Z         }
2021-02-16T13:09:41.1657250Z 
2021-02-16T13:09:41.1657526Z         # (1 unchanged block hidden)
2021-02-16T13:09:41.1657844Z     }
```

But I didn't run terraform apply again.

**Comment by arestarh** [comment]
@manicminer Could you be so kind to take a look ?

**Comment by arestarh** [comment]
What I've noticed:
1. Casing difference for **full_backup_frequency** setting in manual_schedule block
2. Attempting to set default value of 60 minutes for **log_backup_frequency_in_minutes** while Azure reports 0 value (**log_backup_frequency_in_minutes** already set as 60 minutes via Azure Portal)

**Comment by arestarh** [comment]
I've did some more tests:
1. If we set **log_backup_frequency_in_minutes** other than 60 (seems like 60 is a default value) than **terraform apply** will completed successfully and the following **terraform plan** will not detect changes in **azurerm_mssql_virtual_machine** resource.
2.  If we set **log_backup_frequency_in_minutes** as default 60 than we'll have issues described above - the following **terraform plan** will detect changes in **azurerm_mssql_virtual_machine** and subsequent **terraform apply** will fail with error.

**Comment by BazzardTCR** [comment]
In my case, with the creation of a new vm when setting it to 60 terraform "hangs" and I had to abort the run. (after 30 mins).

**Comment by manicminer** [comment]
Hi @arestarh, I'm taking a look at this and trying to reproduce the scenarios you described (thanks for the detailed repro information).

Unfortunately this API is littered with consistency issues which are probably giving rise to the timeouts, and it's broken for cases where auto_backup needs to be disabled, hence the conditional ForceNew. However I'll see if we can improve the logic around this and perhaps adding case insensitivity will help.

**Comment by manicminer** [comment]
@arestarh I've spent some time testing, trying to mimic your workflow as much as possible.

- `auto_backup` isn't a Computed block, so you're correct in this case your Terraform configuration needs to match the auto backup settings for the SQL VM being managed.
- There is a case difference for the `full_backup_frequency` property - when setting this in the Portal it's returned in uppercase - in my testing by itself this didn't create a diff in Terraform due to the suppress.CaseDifference DiffSuppressFunc. I believe this is showing in your diff because of the `log_backup_frequency_in_minutes` property (next point).
- log_backup_frequency_in_minutes: I did observe that the API returns `0` instead of `60`. We can work around this in the provider, I'll link a bugfix PR shortly.

@arestarh The API is inconsistent here which means we have to wait until it returns the correct values after updating the SQL VM. In the event that it never becomes consistent, we continue waiting until the maximum timeout of 60 minutes has elapsed.
